p2p/transport/quic/transport.go

```diff
@@ -38,17 +38,118 @@ var quicDialContext = quic.DialContext // so we can mock it in tests
 
 var HolePunchTimeout = 5 * time.Second
 
+const bufferCap = 5
+
+type bufferEntry struct {
+	handshakes, total uint
+}
+
+type status struct {
+	cycle, pos int
+	check      bool
+}
+
+type accept struct {
+	buffer    []bufferEntry
+	mtx       sync.Mutex
+	duration  time.Duration
+	fraction  float64
+	startTime time.Time
+	current   status
+}
+
+func newAccept(duration time.Duration, fraction float64) *accept {
+	a := new(accept)
+
+	if duration == 0 {
+		duration = 1 * time.Minute
+	}
+	if fraction == 0 {
+		fraction = 0.5
+	}
+
+	a.duration = duration
+	a.fraction = fraction
+	a.startTime = time.Now()
+	a.buffer = make([]bufferEntry, 1, bufferCap)
+	return a
+}
+
+// Call it from inside a mutex to avoid a data race
+func (a *accept) getPosition() (bool, int) {
+	currentCycle := int(time.Since(a.startTime) / a.duration)
+	if currentCycle > a.current.cycle {
+		if len(a.buffer) < cap(a.buffer) {
+			a.buffer = append(a.buffer, bufferEntry{})
```

Comment:
There's an annoying edge case here: There might be minutes during which we don't receive _any_ handshakes (imagine a node that's only receiving new connections very infrequently). We need to clean up old entries in that case.